### DISCUSSION TITLE
[webpack-dev-server] local-ip as webSocketURL hostname

### DISCUSSION THREAD

**Discussion by doougui** [webpack/webpack#15997] [Q&A]
So, I'm developing a Chrome extension to interact with WhatsApp, and, in order to use the `dev` version properly as an extension in my browser, it seems like I need to set my webSocketURL hostname as my ip address. If I don't do this, I get the following error:

![image](https://user-images.githubusercontent.com/44846329/176261422-18022746-4645-4051-b5c0-c09737755c7a.png)

So it seems like It's trying to get the informations (hostname, port, etc) from the browser.

As soon as I change my `webSocketURL hostname` to this:

```js
  devServer: {
    client: {
      webSocketURL: {
        hostname: '179.125.xxx.xx',
        pathname: '/ws',
        protocol: 'ws',
        port: 8080,
      },
    },
  },
```

It works normally.

**My question is: is there any way to set the hostname "dynamically" and not "hardcoded"? Just like we can set the [host](https://webpack.js.org/configuration/dev-server/#devserverhost) property as [local-ip](https://webpack.js.org/configuration/dev-server/#local-ip). Can we do something like that here? Or maybe there's some way to do this in a better way?**

**Comment by alexander-akait** [comment]
Try to set it to `0.0.0.0`

**Reply by doougui** [reply]
First of all, thanks for you answer.

So, about this solution, I've tried it and when I do this, it seems like it tries to get the hostname from the browser again, as you can see in the image below:

![image](https://user-images.githubusercontent.com/44846329/176268302-1f1d8103-a308-4609-a75f-ef7b1e40267b.png)

Code:

```js
module.exports = merge(common, {
  mode: 'development',
  devtool: 'inline-source-map',
  devServer: {
    static: './dist',
    devMiddleware: {
      writeToDisk: true,
    },
    client: {
      webSocketURL: {
        hostname: '0.0.0.0',
        pathname: '/ws',
        protocol: 'ws',
        port: 8080,
      },
    },
    historyApiFallback: true,
  },
});
```

I've also tried to set it to `localhost` and `127.0.0.1`, but I receive the following error when I do this:

![image](https://user-images.githubusercontent.com/44846329/176268785-465a32d8-1815-4738-b284-4a15ff5a268b.png)

In case it helps, this is my full `webpack.common.js`:

```js
const path = require('path');

const { CleanWebpackPlugin } = require('clean-webpack-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const CopyPlugin = require('copy-webpack-plugin');
const TsconfigPathsPlugin = require('tsconfig-paths-webpack-plugin');
const Dotenv = require('dotenv-webpack');

const customPublicPath = path.join(__dirname, './.webpack/customPublicPath');

module.exports = {
  mode: process.env.NODE_ENV || 'development',
  entry: {
    main: [customPublicPath, path.join(__dirname, 'src', 'index.tsx')],
    content: path.join(__dirname, 'src', 'services', 'chrome', 'content', 'index.ts'),
    background: path.join(__dirname, 'src', 'services', 'chrome', 'background', 'index.ts'),
  },
  output: {
    filename: '[name].bundle.js',
    path: path.resolve(__dirname, 'dist'),
    publicPath: './',
  },
  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.jsx'],
    plugins: [new TsconfigPathsPlugin({ configFile: './tsconfig.json' })]
  },
  module: {
    rules: [
      {
        test: /\.ts(x?)$/,
        loader: 'ts-loader',
        exclude: /node_modules/,
      },
      {
        test: /\.svg$/,
        use: ['@svgr/webpack'],
      },
      {
        test: /\.css$/i,
        use: ["style-loader", "css-loader"],
      },
      {
        test: /\.(png|jpg|jpeg|gif)$/i,
        type: 'asset/resource',
      },
    ],
  },
  plugins: [
    new CleanWebpackPlugin(),
    new Dotenv(),
    new HtmlWebpackPlugin({
      template: './public/index.html',
      excludeChunks: ['background', 'content'],
    }),
    new CopyPlugin({
      patterns: [
        {
          from: './public',
          globOptions: {
            ignore: ['**/index.html'],
          },
        },
      ],
    }),
  ],
};
```

and my `webpack.dev.js` (the one i'm running through `webpack-dev-server`):

```js
module.exports = merge(common, {
  mode: 'development',
  devtool: 'inline-source-map',
  devServer: {
    static: './dist',
    devMiddleware: {
      writeToDisk: true,
    },
    client: {
      webSocketURL: {
        hostname: '0.0.0.0',
        pathname: '/ws',
        protocol: 'ws',
        port: 8080,
      },
    },
    historyApiFallback: true,
  },
});
```

**Reply by doougui** [reply]
Just to clarify things, my React application runs as a content script on the browser (the `main` entry in the `webpack.common.js`), so I believe that's the reason that `webpack` returns `web.whatsapp.com` as the "automatic" hostname instead of `localhost` or something like that (if I was developing and testing at http://localhost:8080)

**Reply by alexander-akait** [reply]
hm, if you use domain, why do not use domain in `hostname`? (i.e. `web.whatapp...`)

**Reply by doougui** [reply]
Because `web.whatsapp.com` is just where my application is running at as a chrome extension, not really my domain. The dev server needs to connect the websocket to "my machine" so it "reflects" the changes I do in my code without having to run webpack-dev-server again, re-import the extension on my browser and etc.

The reason why it fails to connect to `wss://web.whatsapp.com:8080/ws` is that it doesn't even exists. `wss://179.125.xxx.xx/ws` (my ip), on the other hand, exists because it's the webpack-dev-server running on my machine.

I hope I managed to explain it well. If I didn't, I can try to set up a demo version later so you or someone who sees this might be able to understand it better :).

**Reply by vankop** [reply]
I don't understand why you need dynamic ip here? you can set your ip in config.. ok I think I get it => you have dynamic ip => you want to set dynamic ip ( your current IP ), but it is not possible to detect it in browser dynamically.
_______

you can do this in webpack config, like
```js
hostname = detectMyIp();
```

**Reply by doougui** [reply]
Yeah, exactly.  That's what I want to do. Since more than one person is going to work on this project the IP must be dynamic.

Unfortunately, it seems like the `detectMyIp` function is not defined in my webpack. Does it require any additional configuration for it to work?

![image](https://user-images.githubusercontent.com/44846329/176424749-8515bdd2-d1b8-4f11-a8b3-6dbd50f9ba2f.png)

```js
const { merge } = require('webpack-merge');
const common = require('./webpack.common.js');

module.exports = merge(common, {
  mode: 'development',
  devtool: 'inline-source-map',
  devServer: {
    static: './dist',
    devMiddleware: {
      writeToDisk: true,
    },
    client: {
      webSocketURL: {
        hostname: detectMyIp(),
        pathname: '/ws',
        protocol: 'ws',
        port: 8080,
      },
    },
    historyApiFallback: true,
  },
});
```

**Reply by vankop** [reply]
I mean - you should implement this ( maybe there is npm package for that )

**Reply by vankop** [reply]
and just use it in your configuration

**Reply by doougui** [reply]
hmmmmm so I think I found a better solution to this. As I said [here](https://github.com/webpack/webpack/discussions/15997#discussioncomment-3042838), I wasn't being able to set my `webSocketURL` to neither `localhost` or `127.0.0.1` because I was receiving a "Invalid Host/Origin header". So I tried setting my `webSocketURL` to `localhost` again and setting my `allowedHosts` option to `all` and it worked normally. My final webpack.dev.js is looking like this now:

```js
module.exports = merge(common, {
  mode: 'development',
  devtool: 'inline-source-map',
  devServer: {
    static: './dist',
    devMiddleware: {
      writeToDisk: true,
    },
    client: {
      webSocketURL: {
        hostname: 'localhost',
        pathname: '/ws',
        protocol: 'ws',
        port: 8080,
      },
    },
    historyApiFallback: true,
    allowedHosts: 'all',
  },
});
```

The only problem now as that HMR and live reload doesn't work that well with Chrome extensions. When I change my code and save it, it auto reloads the web.whatsapp.com page (which is where my extension is running at as a content script) infinitely. This is probably related to this issue: https://github.com/webpack/webpack-dev-server/issues/3861.

What I did was pass the flags `--no-hot` and `--no-live-reload` to my `dev` script.

```js
"dev": "webpack serve --config webpack.dev.js --no-live-reload --no-hot",
```

This way every time I make a change on my code I have to reload my extension on `chrome://extensions` and refresh the page for the changes to reflect on my page. Unfortunately, I think this is the only way to accomplish this now. If there is any way to use `liveReload` and `HMR` with chrome extensions, I would be more than glad to know.

**Reply by alexander-akait** [reply]
You can use `/etc/hosts` and put domaon there and avoid rewrite `webSocketURL`, developers do it, so this problem is not on top due this, using `/etc/hosts` is good practive for development in your case
